### ISSUE TITLE
`torch.onnx.export`'s `dynamic_shape` should take renamed inputs into account

### ISSUE THREAD

**Issue by mqudsi** [pytorch/pytorch#166001]
### 🐛 Describe the bug

Presently, `torch.onnx.export` (with `dynamo=True`) does not take into account renamed inputs and errors out/complains if the explicit labels assigned to the inputs is used in place of the default (e.g. `x` or `x1`) names.

IMHO, the correct behavior here should be that input names are remapped prior to `dynamic_shape` being used/verified, and the renamed inputs should be the only names acceptable to the `dynamic_shape` dict keys. If you wish to _additionally_ support the un-renamed input variables as dict keys, perhaps that's understandable, but the renamed inputs should definitely be supported.

Example:

```python
    onnx_path = "foo.onnx"
    onnx_program = torch.onnx.export(
        model,
        (bgr_tensor_input,),
        input_names=["bgr_image"],
        output_names=["embeddings", "norm"],
        dynamic_shapes={
            "bgr_image": {0: "batch"},
        },
        dynamo=True,
    )
    onnx_program.save(onnx_path)
```

throws the following `torch.onnx._internal.exporter._errors.TorchExportError` exception:

> `<class 'torch._dynamo.exc.UserError'>`: When `dynamic_shapes` is specified as a dict, its top-level keys must be the arg names `['x']` of `inputs`, but here they are `['bgr_image']`. Alternatively, you could also ignore arg names entirely and specify `dynamic_shapes` as a list/tuple matching `inputs`. For more information about this error, see: https://pytorch.org/docs/main/generated/exportdb/index.html#dynamic-shapes-validation

### Versions

```
PyTorch version: 2.9.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 IoT Enterprise LTSC (10.0.26100 64-bit)
GCC version: Could not collect
Clang version: 21.1.2
CMake version: version 3.31.6-msvc6
Libc version: N/A

Python version: 3.13.7 (tags/v3.13.7:bcee1c3, Aug 14 2025, 14:15:11) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26100-SP0
Is CUDA available: False
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080
Nvidia driver version: 581.29
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Name: AMD Ryzen Threadripper 1950X 16-Core Processor
Manufacturer: AuthenticAMD
Family: 107
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 3600
MaxClockSpeed: 3600
L2CacheSize: 8192
L2CacheSpeed: None
Revision: 257

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
```

cc @justinchuby @titaiwangms @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

**Comment by sam-shubham** [comment]
Hi @titaiwangms 👋  

I'd like to work on this issue.

It seems the problem is that `dynamic_shapes` validation uses the original `torch.export` argument names (e.g. `['x']`), while `input_names` renaming happens later at the ONNX level. Because of this, when `dynamic_shapes` uses renamed inputs (like `"bgr_image"`), validation fails before name remapping occurs.

### Proposed Approach

- If `input_names` is provided, build a mapping between original argument names and renamed ONNX names.
- Remap `dynamic_shapes` keys back to original arg names before validation.
- Keep `torch.export` behavior unchanged.

Please let me know if this direction aligns with the intended design.  

Thanks!

**Comment by titaiwangms** [comment]
> Hi [@titaiwangms](https://github.com/titaiwangms) 👋
> 
> I'd like to work on this issue.
> 
> It seems the problem is that `dynamic_shapes` validation uses the original `torch.export` argument names (e.g. `['x']`), while `input_names` renaming happens later at the ONNX level. Because of this, when `dynamic_shapes` uses renamed inputs (like `"bgr_image"`), validation fails before name remapping occurs.
> 
> ### Proposed Approach
> * If `input_names` is provided, build a mapping between original argument names and renamed ONNX names.
> * Remap `dynamic_shapes` keys back to original arg names before validation.
> * Keep `torch.export` behavior unchanged.
> 
> Please let me know if this direction aligns with the intended design.
> 
> Thanks!

@sam-shubham Sure!

Be careful to the data type of `dynamic_shapes` and model args/kwargs. It might be challenging to handle dict/list and optional inputs situations.

**Comment by sam-shubham** [comment]
> > Hi [@titaiwangms](https://github.com/titaiwangms) 👋
> > 
> > I'd like to work on this issue.
> > 
> > It seems the problem is that `dynamic_shapes` validation uses the original `torch.export` argument names (e.g. `['x']`), while `input_names` renaming happens later at the ONNX level. Because of this, when `dynamic_shapes` uses renamed inputs (like `"bgr_image"`), validation fails before name remapping occurs.
> > 
> > ### Proposed Approach
> > * If `input_names` is provided, build a mapping between original argument names and renamed ONNX names.
> > * Remap `dynamic_shapes` keys back to original arg names before validation.
> > * Keep `torch.export` behavior unchanged.
> > 
> > Please let me know if this direction aligns with the intended design.
> > 
> > Thanks!
> 
> @sam-shubham Sure!
> 
> Be careful to the data type of `dynamic_shapes` and model args/kwargs. It might be challenging to handle dict/list and optional inputs situations.

Got it, thanks @titaiwangms. I’ll keep you posted on the updates. Could you please assign this to me?
